hoomd/md/ForceComposite.cc

```diff
@@ -1048,6 +1054,47 @@ void ForceComposite::updateCompositeParticles(uint64_t timestep)
         }
     }
 
+void ForceComposite::findRigidCenters()
+    {
+    ArrayHandle<unsigned int> h_tag(m_pdata->getTags(), access_location::host, access_mode::read);
+    ArrayHandle<unsigned int> h_rtag(m_pdata->getRTags(), access_location::host, access_mode::read);
+    ArrayHandle<unsigned int> h_body(m_pdata->getBodies(),
+                                     access_location::host,
+                                     access_mode::read);
+
+    m_rigid_center.resize(m_pdata->getN() + m_pdata->getNGhosts());
+    m_rigid_center.zeroFill();
+    m_lookup_center.resize(m_pdata->getN() + m_pdata->getNGhosts());
+    m_lookup_center.zeroFill();
+    ArrayHandle<unsigned int> h_rigid_center(m_rigid_center,
+                                             access_location::host,
+                                             access_mode::overwrite);
+    ArrayHandle<unsigned int> h_lookup_center(m_lookup_center,
+                                              access_location::host,
+                                              access_mode::overwrite);
+    unsigned int num_centers = 0;
+    unsigned int n_particles_local = m_pdata->getN() + m_pdata->getNGhosts();
+    for (unsigned int idx = 0; idx < n_particles_local; idx++)
+        {
+        // check that the particle is in a rigid body and a central particle
+        h_lookup_center.data[idx] = -1;
+        const unsigned int central_tag = h_body.data[idx];
+        if (central_tag >= MIN_FLOPPY)
+            {
+            continue;
+            }
+        const unsigned int central_idx = h_rtag.data[central_tag];
+        h_lookup_center.data[idx] = central_idx;
+        if (central_idx != idx)
+            {
+            continue;
+            }
+        h_rigid_center.data[num_centers] = central_idx;
+        num_centers += 1;
```

Comment:
`lookup_center` should be assigned `NO_BODY` by default. Then, you can change these loops to have if rather than if not behavior:

```suggestion
        unsigned int lookup_center = NO_BODY;

        // if particle is in a rigid body, find its center
        const unsigned int central_tag = h_body.data[idx];
        if (central_tag < MIN_FLOPPY)
            {
            lookup_center = h_rtag.data[central_tag];
            // also record rigid particles that are centers
            if (lookup_center == idx)
                {
                h_rigid_center.data[num_centers++] = lookup_center;
                }
            }

        h_lookup_center.data[idx] = lookup_center;
```

Reply:
does the `num_centers++` in your line 1089 increase num_centers by 1? To me it doesn't look like it does.

Reply:
Yes, it's called postfix increment. `i++` can be thought of as "return the current value of i, then increment i by 1".

Contrast with prefix increment: `++i` can be thought of as "increment i by 1, then return the new value of i"